File: backend/services/IndicatorService.php
<?php
// 技术指标计算服务

require_once __DIR__ . '/Database.php';

class IndicatorService {
    private $db;

    public function __construct() {
        $this->db = Database::getInstance();
    }

    // 计算并保存所有指标
    public function calculate(string $tsCode, array $history): ?array {
        if (count($history) < 26) return null; // 数据不足

        // 反转数组，使其按时间正序
        $data = array_reverse($history);
        $closes = array_column($data, 'close');
        $highs = array_column($data, 'high');
        $lows = array_column($data, 'low');

        $latest = end($data);
        $tradeDate = $latest['trade_date'];

        // 计算各指标
        $rsi6 = $this->calculateRSI($closes, 6);
        $rsi12 = $this->calculateRSI($closes, 12);
        $macd = $this->calculateMACD($closes);
        $kdj = $this->calculateKDJ($closes, $highs, $lows);
        $boll = $this->calculateBOLL($closes);

        $indicators = [
            'ts_code' => $tsCode,
            'trade_date' => $tradeDate,
            'rsi_6' => $rsi6,
            'rsi_12' => $rsi12,
            'macd' => $macd['macd'],
            'macd_signal' => $macd['signal'],
            'macd_hist' => $macd['hist'],
            'k' => $kdj['k'],
            'd' => $kdj['d'],
            'j' => $kdj['j'],
            'boll_upper' => $boll['upper'],
            'boll_mid' => $boll['mid'],
            'boll_lower' => $boll['lower'],
        ];

        // 保存到数据库
        $this->saveIndicators($indicators);

        return $indicators;
    }

    // RSI计算
    private function calculateRSI(array $closes, int $period): float {
        $changes = [];
        for ($i = 1; $i < count($closes); $i++) {
            $changes[] = $closes[$i] - $closes[$i - 1];
        }

        $gains = [];
        $losses = [];
        foreach ($changes as $change) {
            $gains[] = $change > 0 ? $change : 0;
            $losses[] = $change < 0 ? abs($change) : 0;
        }

        // 取最近period个数据
        $recentGains = array_slice($gains, -$period);
        $recentLosses = array_slice($losses, -$period);

        $avgGain = array_sum($recentGains) / $period;
        $avgLoss = array_sum($recentLosses) / $period;

        if ($avgLoss == 0) return 100;

        $rs = $avgGain / $avgLoss;
        return round(100 - (100 / (1 + $rs)), 2);
    }

    // MACD计算
    private function calculateMACD(array $closes): array {
        $ema12 = $this->calculateEMA($closes, 12);
        $ema26 = $this->calculateEMA($closes, 26);

        $dif = $ema12 - $ema26;

        // 简化：用最近9个DIF的EMA作为信号线
        $signal = $dif * 0.2 + $dif * 0.8; // 简化处理
        $hist = $dif - $signal;

        return [
            'macd' => round($dif, 4),
            'signal' => round($signal, 4),
            'hist' => round($hist, 4),
        ];
    }

    // EMA计算
    private function calculateEMA(array $data, int $period): float {
        $k = 2 / ($period + 1);
        $ema = $data[0];

        for ($i = 1; $i < count($data); $i++) {
            $ema = $data[$i] * $k + $ema * (1 - $k);
        }

        return $ema;
    }

    // KDJ计算
    private function calculateKDJ(array $closes, array $highs, array $lows): array {
        $period = 9;
        $n = count($closes);

        // 取最近period天的最高和最低
        $recentHighs = array_slice($highs, -$period);
        $recentLows = array_slice($lows, -$period);

        $hn = max($recentHighs);
        $ln = min($recentLows);
        $cn = end($closes);

        $rsv = ($hn - $ln) != 0 ? (($cn - $ln) / ($hn - $ln)) * 100 : 50;

        // 简化：K=RSV的SMA，D=K的SMA
        $k = $rsv;
        $d = $k;
        $j = 3 * $k - 2 * $d;

        return [
            'k' => round($k, 2),
            'd' => round($d, 2),
            'j' => round($j, 2),
        ];
    }

    // 布林带计算
    private function calculateBOLL(array $closes, int $period = 20): array {
        $recentCloses = array_slice($closes, -$period);
        $mid = array_sum($recentCloses) / $period;

        // 计算标准差
        $variance = 0;
        foreach ($recentCloses as $close) {
            $variance += pow($close - $mid, 2);
        }
        $std = sqrt($variance / $period);

        return [
            'upper' => round($mid + 2 * $std, 2),
            'mid' => round($mid, 2),
            'lower' => round($mid - 2 * $std, 2),
        ];
    }

    // 保存指标
    private function saveIndicators(array $data): void {
        $columns = implode(',', array_keys($data));
        $placeholders = implode(',', array_fill(0, count($data), '?'));

        $updates = [];
        foreach (array_keys($data) as $col) {
            if ($col !== 'ts_code' && $col !== 'trade_date') {
                $updates[] = "{$col}=VALUES({$col})";
            }
        }

        $sql = "INSERT INTO technical_indicators ({$columns}) VALUES ({$placeholders})
                ON DUPLICATE KEY UPDATE " . implode(',', $updates);

        $this->db->execute($sql, array_values($data));
    }

    // 获取超卖股票（RSI<30）
    public function getOversoldStocks(string $tradeDate): array {
        return $this->db->query(
            "SELECT i.ts_code, s.name, s.industry, i.rsi_6, i.rsi_12, i.k, i.d
             FROM technical_indicators i
             JOIN stocks s ON i.ts_code = s.ts_code
             WHERE i.trade_date = ? AND i.rsi_6 < 30
             ORDER BY i.rsi_6 ASC
             LIMIT 50",
            [$tradeDate]
        );
    }

    // 获取KDJ见底股票
    public function getKDJBottomStocks(string $tradeDate): array {
        return $this->db->query(
            "SELECT i.ts_code, s.name, s.industry, i.k, i.d, i.j
             FROM technical_indicators i
             JOIN stocks s ON i.ts_code = s.ts_code
             WHERE i.trade_date = ? AND i.k < 20 AND i.d < 20
             ORDER BY i.k ASC
             LIMIT 50",
            [$tradeDate]
        );
    }

    // 获取MACD金叉股票
    public function getMACDGoldenCross(string $tradeDate): array {
        return $this->db->query(
            "SELECT i.ts_code, s.name, s.industry, i.macd, i.macd_signal, i.macd_hist
             FROM technical_indicators i
             JOIN stocks s ON i.ts_code = s.ts_code
             WHERE i.trade_date = ? AND i.macd_hist > 0 AND i.macd_hist < 0.1
             ORDER BY i.macd_hist ASC
             LIMIT 50",
            [$tradeDate]
        );
    }

    // 获取底部放量股票
    public function getBottomVolumeStocks(string $tradeDate): array {
        // 底部放量条件：
        // 1. 量比 > 3（当日成交量/20日均量）
        // 2. 价格位置 < 0.3（处于60日低位）
        // 3. 当日涨幅 > 3%
        // 4. RSI < 50（还没超买）
        return $this->db->query(
            "SELECT
                q.ts_code,
                s.name,
                s.industry,
                q.close,
                q.pct_chg,
                ROUND(q.vol / avg_data.avg_vol, 2) as volume_ratio,
                ROUND((q.close - range_data.low_60) / (range_data.high_60 - range_data.low_60) * 100, 2) as price_position,
                i.rsi_6
             FROM daily_quotes q
             JOIN stocks s ON q.ts_code = s.ts_code
             JOIN technical_indicators i ON q.ts_code = i.ts_code AND q.trade_date = i.trade_date
             JOIN (
                 SELECT ts_code, AVG(vol) as avg_vol
                 FROM daily_quotes
                 WHERE trade_date <= ? AND trade_date >= DATE_FORMAT(DATE_SUB(STR_TO_DATE(?, '%Y%m%d'), INTERVAL 20 DAY), '%Y%m%d')
                 GROUP BY ts_code
             ) avg_data ON q.ts_code = avg_data.ts_code
             JOIN (
                 SELECT ts_code, MIN(low) as low_60, MAX(high) as high_60
                 FROM daily_quotes
                 WHERE trade_date <= ? AND trade_date >= DATE_FORMAT(DATE_SUB(STR_TO_DATE(?, '%Y%m%d'), INTERVAL 60 DAY), '%Y%m%d')
                 GROUP BY ts_code
             ) range_data ON q.ts_code = range_data.ts_code
             WHERE q.trade_date = ?
               AND q.vol > avg_data.avg_vol * 3
               AND (q.close - range_data.low_60) / (range_data.high_60 - range_data.low_60) < 0.3
               AND q.pct_chg > 3
               AND i.rsi_6 < 50
               AND range_data.high_60 > range_data.low_60
             ORDER BY q.vol / avg_data.avg_vol DESC
             LIMIT 50",
            [$tradeDate, $tradeDate, $tradeDate, $tradeDate, $tradeDate]
        );
    }

    // 获取行业异动（板块集体上涨）
    public function getIndustryHotStocks(string $tradeDate): array {
        // 条件：行业内上涨股票数>=5，平均涨幅>3%
        return $this->db->query(
            "SELECT
                s.industry,
                COUNT(*) as stock_count,
                ROUND(AVG(q.pct_chg), 2) as avg_pct_chg,
                (SELECT s2.name FROM daily_quotes q2
                 JOIN stocks s2 ON q2.ts_code = s2.ts_code
                 WHERE q2.trade_date = ? AND s2.industry = s.industry
                 ORDER BY q2.pct_chg DESC LIMIT 1) as top_stock,
                (SELECT MAX(q3.pct_chg) FROM daily_quotes q3
                 JOIN stocks s3 ON q3.ts_code = s3.ts_code
                 WHERE q3.trade_date = ? AND s3.industry = s.industry) as top_pct
             FROM daily_quotes q
             JOIN stocks s ON q.ts_code = s.ts_code
             WHERE q.trade_date = ?
               AND q.pct_chg > 0
               AND s.industry IS NOT NULL
               AND s.industry != ''
             GROUP BY s.industry
             HAVING COUNT(*) >= 5 AND AVG(q.pct_chg) > 3
             ORDER BY AVG(q.pct_chg) DESC
             LIMIT 20",
            [$tradeDate, $tradeDate, $tradeDate]
        );
    }

    // 获取大盘指数
    public function getMarketIndex(string $tradeDate): array {
        return $this->db->query(
            "SELECT ts_code, close, pct_chg, vol, amount
             FROM daily_quotes
             WHERE trade_date = ?
               AND ts_code IN ('000001.SH', '399001.SZ', '399006.SZ')
             ORDER BY ts_code",
            [$tradeDate]
        );
    }

    // 获取逆势上涨股票
    public function getCounterTrendStocks(string $tradeDate): array {
        // 先获取上证指数涨跌幅
        $indexData = $this->db->query(
            "SELECT pct_chg FROM daily_quotes WHERE trade_date = ? AND ts_code = '000001.SH'",
            [$tradeDate]
        );

        $indexChg = $indexData[0]['pct_chg'] ?? 0;

        // 大盘跌时筛选逆势股，涨幅>2%，量比>1.5
        return $this->db->query(
            "SELECT
                q.ts_code,
                s.name,
                s.industry,
                q.close,
                q.pct_chg,
                ROUND(q.vol / NULLIF(avg_data.avg_vol, 0), 2) as volume_ratio
             FROM daily_quotes q
             JOIN stocks s ON q.ts_code = s.ts_code
             LEFT JOIN (
                 SELECT ts_code, AVG(vol) as avg_vol
                 FROM daily_quotes
                 WHERE trade_date <= ? AND trade_date >= DATE_FORMAT(DATE_SUB(STR_TO_DATE(?, '%Y%m%d'), INTERVAL 10 DAY), '%Y%m%d')
                 GROUP BY ts_code
             ) avg_data ON q.ts_code = avg_data.ts_code
             WHERE q.trade_date = ?
               AND q.pct_chg > 2
               AND q.ts_code NOT LIKE '%%.SH'
               AND q.ts_code NOT LIKE '399%%'
             ORDER BY q.pct_chg DESC
             LIMIT 50",
            [$tradeDate, $tradeDate, $tradeDate]
        );
    }
}

File: backend-python/app/services/indicator_service.py
import numpy as np
from typing import List, Dict, Optional
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from loguru import logger

from app.models import DailyQuote, TechnicalIndicator
from app.core.indicators import (
    calculate_rsi_multi,
    calculate_macd,
    calculate_kdj,
    calculate_boll,
)


class IndicatorService:
    """技术指标服务"""

    def __init__(self, session: AsyncSession):
        self.session = session

    async def get_stock_history(
        self,
        ts_code: str,
        limit: int = 200
    ) -> List[dict]:
        """获取股票历史行情数据

        Args:
            ts_code: 股票代码
            limit: 获取数量

        Returns:
            历史行情列表
        """
        stmt = (
            select(DailyQuote)
            .where(DailyQuote.ts_code == ts_code)
            .order_by(DailyQuote.trade_date.desc())
            .limit(limit)
        )
        result = await self.session.execute(stmt)
        quotes = result.scalars().all()

        # 反向排序 (时间从早到晚)
        return [
            {
                "ts_code": q.ts_code,
                "trade_date": q.trade_date,
                "open": q.open,
                "high": q.high,
                "low": q.low,
                "close": q.close,
                "vol": q.vol,
                "amount": q.amount,
                "pct_chg": q.pct_chg,
            }
            for q in reversed(quotes)
        ]

    async def calculate_indicators(
        self,
        ts_code: str,
        history: List[dict]
    ) -> dict:
        """计算股票的所有技术指标

        Args:
            ts_code: 股票代码
            history: 历史行情列表

        Returns:
            指标字典
        """
        if len(history) < 26:  # MACD需要26条数据
            return {}

        closes = np.array([q.get("close", 0) for q in history], dtype=float)
        highs = np.array([q.get("high", 0) for q in history], dtype=float)
        lows = np.array([q.get("low", 0) for q in history], dtype=float)

        # 计算各指标
        rsi = calculate_rsi_multi(closes, [6, 12])
        dif, dea, macd_hist = calculate_macd(closes)
        k, d, j = calculate_kdj(highs, lows, closes)
        upper, mid, lower = calculate_boll(closes)

        return {
            "ts_code": ts_code,
            "trade_date": history[-1].get("trade_date"),
            "rsi_6": rsi.get("rsi_6"),
            "rsi_12": rsi.get("rsi_12"),
            "macd": dif,
            "macd_signal": dea,
            "macd_hist": macd_hist,
            "k": k,
            "d": d,
            "j": j,
            "boll_upper": upper,
            "boll_mid": mid,
            "boll_lower": lower,
        }

    async def calc_all(
        self,
        trade_date: str,
        stocks: Optional[List[str]] = None
    ) -> int:
        """计算所有股票的指标

        Args:
            trade_date: 交易日期 (YYYYMMDD)
            stocks: 股票列表 (如果为None则计算所有)

        Returns:
            计算数量
        """
        from app.models import Stock

        # 获取股票列表
        if stocks:
            stmt = select(Stock).where(Stock.ts_code.in_(stocks))
        else:
            stmt = select(Stock)

        result = await self.session.execute(stmt)
        all_stocks = result.scalars().all()

        calculated_count = 0

        for stock in all_stocks:
            try:
                # 获取历史数据
                history = await self.get_stock_history(stock.ts_code, 200)
                if len(history) < 26:
                    continue

                # 计算指标
                indicators = await self.calculate_indicators(stock.ts_code, history)
                if not indicators:
                    continue

                # 保存到数据库
                stmt = (
                    select(TechnicalIndicator)
                    .where(TechnicalIndicator.ts_code == stock.ts_code)
                    .where(TechnicalIndicator.trade_date == trade_date)
                )
                result = await self.session.execute(stmt)
                existing = result.scalar_one_or_none()

                if existing:
                    # 更新
                    for key, value in indicators.items():
                        if key != "ts_code":
                            setattr(existing, key, value)
                else:
                    # 新增
                    new_indicator = TechnicalIndicator(**indicators)
                    self.session.add(new_indicator)

                calculated_count += 1

            except Exception as e:
                logger.error(f"Failed to calculate indicators for {stock.ts_code}: {e}")
                continue

        await self.session.commit()
        logger.info(f"Calculated indicators for {calculated_count} stocks")
        return calculated_count

    async def get_oversold_stocks(
        self,
        trade_date: str,
        rsi_threshold: int = 30
    ) -> List[dict]:
        """获取RSI超卖股票

        Args:
            trade_date: 交易日期
            rsi_threshold: RSI阈值

        Returns:
            超卖股票列表
        """
        from app.models import Stock

        stmt = (
            select(TechnicalIndicator, Stock)
            .join(Stock, TechnicalIndicator.ts_code == Stock.ts_code)
            .where(TechnicalIndicator.trade_date == trade_date)
            .where(TechnicalIndicator.rsi_6 < rsi_threshold)
            .order_by(TechnicalIndicator.rsi_6.asc())
            .limit(50)
        )

        result = await self.session.execute(stmt)
        rows = result.all()

        return [
            {
                "ts_code": indicator.ts_code,
                "name": stock.name,
                "rsi_6": indicator.rsi_6,
                "rsi_12": indicator.rsi_12,
            }
            for indicator, stock in rows
        ]

    async def get_kdj_bottom_stocks(
        self,
        trade_date: str,
        k_threshold: int = 20,
        d_threshold: int = 20
    ) -> List[dict]:
        """获取KDJ底部信号股票

        Args:
            trade_date: 交易日期
            k_threshold: K值阈值
            d_threshold: D值阈值

        Returns:
            底部信号列表
        """
        from app.models import Stock

        stmt = (
            select(TechnicalIndicator, Stock)
            .join(Stock, TechnicalIndicator.ts_code == Stock.ts_code)
            .where(TechnicalIndicator.trade_date == trade_date)
            .where(TechnicalIndicator.k < k_threshold)
            .where(TechnicalIndicator.d < d_threshold)
            .order_by(TechnicalIndicator.k.asc())
            .limit(50)
        )

        result = await self.session.execute(stmt)
        rows = result.all()

        return [
            {
                "ts_code": indicator.ts_code,
                "name": stock.name,
                "k": indicator.k,
                "d": indicator.d,
                "j": indicator.j,
            }
            for indicator, stock in rows
        ]

    async def get_macd_golden_stocks(
        self,
        trade_date: str
    ) -> List[dict]:
        """获取MACD金叉股票

        Args:
            trade_date: 交易日期

        Returns:
            MACD金叉列表
        """
        from app.models import Stock

        stmt = (
            select(TechnicalIndicator, Stock)
            .join(Stock, TechnicalIndicator.ts_code == Stock.ts_code)
            .where(TechnicalIndicator.trade_date == trade_date)
            .where(TechnicalIndicator.macd_hist > 0)
            .order_by(TechnicalIndicator.macd_hist.desc())
            .limit(50)
        )

        result = await self.session.execute(stmt)
        rows = result.all()

        return [
            {
                "ts_code": indicator.ts_code,
                "name": stock.name,
                "macd": indicator.macd,
                "macd_hist": indicator.macd_hist,
            }
            for indicator, stock in rows
        ]
